### PR TITLE
Add a unit test for native_find_rsc

### PR DESCRIPTION
**Pull request by clumens**
I am especially interested in the clone portion of this.

Also, this really skews the coverage numbers because of all the stuff called in the setup function.  That technically gets run during "make check", which means it's now covered by a test case.  Of course, that's not true - we're not actually testing any of that stuff, we're assuming it works in order to test something else.  So I might need to figure out some way to work around that.

### CODE REVIEW

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    assert_null(native_find_rsc(inactive_group, "inactive-group", cluster01, pe_find_clone));
+
+    /* Fails because none of inactive-group's children are running. */
+    assert_null(native_find_rsc(inactive_group, "inactive-group", cluster01, pe_find_current));
+    assert_null(native_find_rsc(inactive_group, "inactive-group", cluster02, pe_find_current));
+
+    /* Passes because of flags. */
+    assert_ptr_equal(inactive_group, native_find_rsc(inactive_group, "inactive-group", cluster01, pe_find_inactive));
+    /* Passes because of flags. */
+    assert_ptr_equal(inactive_group, native_find_rsc(inactive_group, "inactive-group", cluster02, pe_find_inactive));
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:177]: for completeness, a NULL node with pe_find_inactive should work too

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+            ping_clone = rsc;
+            break;
+        }
+    }
+
+    assert_non_null(ping_clone);
+
+    /* Passes because NULL was passed for node, regardless of flags. */
+    assert_ptr_equal(ping_clone, native_find_rsc(ping_clone, "ping-clone", NULL, 0));
+    assert_ptr_equal(ping_clone, native_find_rsc(ping_clone, "ping-clone", NULL, pe_find_current));
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:278]: pe_find_clone with NULL node is worthwhile too

**Review by kgaillot — COMMENTED** [review]
> lso, this really skews the coverage numbers because of all the stuff called in the setup function. That technically gets run during "make check", which means it's now covered by a test case. Of course, that's not true - we're not actually testing any of that stuff, we're assuming it works in order to test something else. So I might need to figure out some way to work around that.

Obviously the workaround is to unit test all of it ;)

This is a great start. To make it official we also need to test cloned groups, bundles, pe_find_any, pe_find_anon, and pe_find_renamed. A separate unit test for pe_base_name_eq() would make the coverage more correct.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+pe_node_t *cluster01, *cluster02;
+
+static int
+setup(void **state) {
+    char *path = NULL;
+
+    crm_xml_init();
+
+    path = crm_strdup_printf("%s/crm_mon.xml", getenv("CTS_CLI_DIR"));
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:34]: Fixed.

**Comment by clumens** [comment]
> Obviously the workaround is to unit test all of it ;)

Oh man, there's just so much of it too.

Just adding this one test case bumps our line coverage from something like 1.9% to 6.4%.  It's crazy how much stuff gets run by calling pe_new_working_set(), cluster_status(), and pe_find_node().

> This is a great start. To make it official we also need to test cloned groups, bundles, pe_find_any, pe_find_anon, and pe_find_renamed. A separate unit test for pe_base_name_eq() would make the coverage more correct.

Yeah, my one commit message says I still need to test cloned groups, bundles, and inactive clones.  Clone testing kind of falls apart too, because I'm not sure what's expected for all the various combinations of flags and resource names.  Basically, this PR currently contains everything I'm pretty sure about, and then where my understanding drops off, the tests stop.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        }
+    }
+
+    assert_non_null(dummy);
+
+    /* Passes because NULL was passed for node, regardless of flags. */
+    assert_ptr_equal(dummy, native_find_rsc(dummy, "dummy", NULL, 0));
+    assert_ptr_equal(dummy, native_find_rsc(dummy, "dummy", NULL, pe_find_current));
+
+    /* Fails because resource is not a clone. */
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:98]: Fixed comments.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 1 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:177]: Added.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 2 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:278]: Added.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+static void
+bad_args(void **state) {
+    pe_resource_t *rsc = (pe_resource_t *) g_list_first(data_set->resources)->data;
+
+    assert_non_null(rsc);
+
+    assert_null(native_find_rsc(NULL, "dummy", NULL, 0));
+    assert_null(native_find_rsc(rsc, NULL, NULL, 0));
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:75]: Added both.

**Comment by clumens** [comment]
Pushed with updates for most comments, additional clone testing, inactive clone testing, and the beginnings of bundles.  I'm not sure how to deal with pe_find_anon and pe_find_renamed.  I'm also not sure what to do about replicas inside bundles, given that is hidden information.

**Comment by clumens** [comment]
Updated with tests for pe_find_anon, pe_find_renamed, and bundles.  I'm going to remove the WIP from this since I think it's good enough to consider.

I'm still looking for a way to not have all the setup code be considered in the coverage numbers, since that's going to make it harder for me to see what still needs tests.  I'd also be interested in any ways to make the test case a little less verbose - grabbing all the resources it needs in the loop in setup, perhaps, or maybe just reducing some of the stuff I am testing that might not need to be there.

Also, maybe I should convert one of the clone resources in cts/cli/crm_mon.xml to be a unique clone just so we have something testing that they get displayed as we expect.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+            promotable_1 = rsc;
+        }
+    }
+
+    assert_non_null(promotable_0);
+    assert_non_null(promotable_1);
+
+    /* Passes because pe_find_renamed means the base name matches clone_name. */
+    assert_ptr_equal(promotable_0, native_find_rsc(promotable_0, "promotable-rsc", NULL, pe_find_renamed));
+    assert_ptr_equal(promotable_1, native_find_rsc(promotable_1, "promotable-rsc", NULL, pe_find_renamed));
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:414]: I wonder if pe_find_renamed is effectively any different from pe_find_clone. We definitely have too many find options, and I'd love to ditch clone_name someday if possible. I'm not sure whether clone_name can be different from ID(rsc->xml); possibly in a corner case like the unique-to-anonymous conversion.

BTW I just noticed that pe_find_clone incorrectly uses pcmk__str_casei, it should be pcmk__str_none -- maybe do that as a commit before the unit test.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+#include <cmocka.h>
+
+#include <crm/common/xml.h>
+#include <crm/pengine/internal.h>
+#include <crm/pengine/status.h>
+#include <crm/pengine/pe_types.h>
+
+/* Needed to access replicas inside a bundle. */
+#define PE__VARIANT_BUNDLE 1
+#include <lib/pengine/variant.h>
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:25]: Unrelated to this PR, just FYI, my long-term plan is to add enough (internal-only) accessors that the variant.h stuff can be local to each variant .c file.

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 8]
```diff
@@ -0,0 +1,683 @@
...
+    child_0 = replica_0->child;
+    container_0 = replica_0->container;
+    remote_0 = replica_0->remote;
+
+    assert_non_null(ip_0);
+    assert_non_null(child_0);
+    assert_non_null(container_0);
+    assert_non_null(remote_0);
+
+    /* Passes beacuse NULL was passed for node, regardless of flags. */
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:459]: spp

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 9]
```diff
@@ -0,0 +1,683 @@
...
+        cmocka_unit_test(inactive_group_member_rsc),
+        cmocka_unit_test(clone_rsc),
+        cmocka_unit_test(inactive_clone_rsc),
+        cmocka_unit_test(clone_instance_rsc),
+        cmocka_unit_test(renamed_rsc),
+        cmocka_unit_test(bundle_rsc),
+        cmocka_unit_test(bundle_replica_rsc),
+        cmocka_unit_test(clone_group_rsc),
+        cmocka_unit_test(clone_group_instance_rsc),
+        cmocka_unit_test(clone_group_member_rsc),
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:678]: Add a todo comment that we should also test finding on allocated node (passing a node without pe_find_current, after scheduling has been done, for a resource that's starting, stopping, or moving). It's not worth the trouble right now ...

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 10]
```diff
@@ -0,0 +1,683 @@
...
+        if (strcmp(rsc->id, "mysql-group:0") == 0) {
+            for (GList *iter2 = rsc->children; iter2 != NULL; iter2 = iter2->next) {
+                pe_resource_t *child = (pe_resource_t *) iter2->data;
+
+                if (strcmp(child->id, "mysql-proxy:0") == 0) {
+                    mysql_proxy = child;
+                    break;
+                }
+            }
+        }
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:639]: You probably want to break at the end of the if block as well

File: lib/pengine/tests/native/native_find_rsc_test.c [hunk 11]
```diff
@@ -0,0 +1,683 @@
...
+
+    /* Fails because resource is not a clone (nor cloned). */
+    assert_null(native_find_rsc(dummy, "dummy", NULL, pe_find_clone));
+    assert_null(native_find_rsc(dummy, "dummy", cluster02, pe_find_clone));
+
+    /* Fails because dummy is not running on cluster01, even with the right flags. */
+    assert_null(native_find_rsc(dummy, "dummy", cluster01, pe_find_current));
+
+    /* Fails because pe_find_current is required if a node is given. */
+    assert_null(native_find_rsc(dummy, "dummy", cluster02, 0));
```
Comment (kgaillot) [inline lib/pengine/tests/native/native_find_rsc_test.c:151]: To clarify further, this fails only because we didn't call pcmk__schedule_actions(). If we had done that, then (in this case) dummy is staying on cluster02, so the allocated node would match. But since we only called cluster_status(), we don't have any allocated node information.

To put it another way -- we could call cluster_status(), run this test as is, then call pcmk__schedule_actions(), then repeat this test, and it would pass on the second call.

Alternatively, we could just call pcmk__schedule_actions() instead of cluster_status(), and be able to test current vs allocated nodes.

**Review by kgaillot — COMMENTED** [review]
Whew! Talk about brain-straining. Looks good, let me know when you're ready to merge.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 6 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:414]: Fixed the case comparison.  It didn't change the test output.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 7 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:25]: That would be really nice to have.  It comes up for me every now and then, and having to do this define/include trick feels dirty.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 10 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:639]: This is handled by the `mysql_proxy == NULL` condition in the for loop, but that's probably too subtle.  I kind of hate adding extra end conditions into the for loop.  I don't know why I did it here.  I'll change it to a break.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 9 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:678]: Added.

File: lib/pengine/tests/native/native_find_rsc_test.c (hunk 11 above)
Comment (clumens) [reply lib/pengine/tests/native/native_find_rsc_test.c:151]: I don't think it would be difficult to test both before and after running pcmk__schedule_actions, just verbose.  I think what we could do is pass some state variable to each test - say, a bool that's false if these should pass.  We'd run the tests, call pcmk__schedule_actions, and then run them again with that bool now set to true.  The verbose part would be conditionalizing all these tests that can later pass.

**Comment by clumens** [comment]
> Whew! Talk about brain-straining. Looks good, let me know when you're ready to merge.

I should add a pe_base_name_eq test first.
